pageserver/src/tenant.rs

```diff
@@ -706,30 +706,37 @@ impl Tenant {
             );
         }
         // Wait for all the download tasks to complete & collect results.
-        let mut remote_clients = HashMap::new();
-        let mut index_parts = HashMap::new();
+        let mut remote_index_and_client = HashMap::new();
         let mut timeline_ancestors = HashMap::new();
         while let Some(result) = part_downloads.join_next().await {
             // NB: we already added timeline_id as context to the error
             let result: Result<_, anyhow::Error> = result.context("joinset task join")?;
             let (timeline_id, client, index_part, remote_metadata) = result?;
             debug!("successfully downloaded index part for timeline {timeline_id}");
             timeline_ancestors.insert(timeline_id, remote_metadata);
-            index_parts.insert(timeline_id, index_part);
-            remote_clients.insert(timeline_id, client);
+            remote_index_and_client.insert(timeline_id, (index_part, client));
         }
 
         // For every timeline, download the metadata file, scan the local directory,
         // and build a layer map that contains an entry for each remote and local
         // layer file.
         let sorted_timelines = tree_sort_timelines(timeline_ancestors)?;
         for (timeline_id, remote_metadata) in sorted_timelines {
+            let (index_part, remote_client) = remote_index_and_client
+                .remove(&timeline_id)
+                .expect("just put it in above");
+
+            if index_part.is_deleted {
+                info!("timeline {} is deleted, skipping", timeline_id);
```

Comment:
Local data is expected to be cleared in delete, which should be retried in case of a failure. (good idea to write such a test, will do)

In the next step I would implement deletion from remote, which can take long time, so it will make sense to respond to delete with accepted and continue clean up operations in the background taking restarts into account

Reply:
Resolving this one, feel free to reopen

Reply:
Added a note about test with retrying delete operation [here](https://github.com/neondatabase/neon/issues/3667)

Related: https://github.com/neondatabase/neon/issues/3478

Reply:
> I'd still mention that we did not find the timeline in the remote storage, hence skipping: it could be present locally, if I've got the code right?

Sorry apparently missed the question. Yes, you're right. Thinking if we should allow for delete retries to succeed for such timeline. I e we updated the file on remote, then didnt finish local cleanup and restarted. After that we either need to delete local files (as you mentioned) or allow for deletion to be retried so this can still be done. Is it what you've thought about?

Reply:
Pretty much it, and I've thought deleting at least the local files could be good to do in the current PR, but no strong opinion on the remote files deletions, that's up for future PRs maybe?
I'd lean to delete request retries as a simpler approach, but that will leak us some data eventually.

Reply:
Yep, deleting remote stuff is for sure for different PR, I'm looking into deleting local files, this looks simpler than supporting delete for such timelines

Reply:
Thinking about the order at which local files should be deleted.

1. remove metadata first then remove local layers

in case there is no metadata timeline cannot be loaded -> safe
in case we see timeline directory without metadata it should be deleted 

2. remove layer files first then remove metadata

Timeline still can be loaded. If remote storage is disabled in the config during such restart we'll get broken timeline

In this case during load we get to the step of checking with remote storage, then see the deleted flag set and then continue with local deletion.

*** 

Currently delete_timeline uses `remove_dir_all` so order is not specified thus we can see arbitrary case appearing.

Another option would be to use `remove_dir_all` here as well and clean it up in all places in a follow up PR. @SomeoneToIgnore @koivunej thoughts?

Reply:
Agreed the `remove_dir_all` is probably fine for now. Had missed the ping.